consensus/state_processing/src/per_epoch_processing/epoch_processing_summary.rs

```diff
@@ -0,0 +1,280 @@
+            EpochProcessingSummary::Altair {
+                participation_cache,
+                ..
+            } => participation_cache.is_current_epoch_timely_target_attester(val_index),
+        }
+    }
+
+    /// Returns the sum of the effective balance of all validators in the previous epoch who
+    /// included an attestation that matched the target.
+    pub fn previous_epoch_target_attesting_balance(&self) -> Result<u64, ParticipationCacheError> {
+        match self {
+            EpochProcessingSummary::Base { total_balances, .. } => {
+                Ok(total_balances.previous_epoch_target_attesters())
+            }
+            EpochProcessingSummary::Altair {
+                participation_cache,
+            } => participation_cache.previous_epoch_target_attesting_balance(),
+        }
+    }
+
+    /// Returns the sum of the effective balance of all validators in the previous epoch who
+    /// included an attestation that matched the head.
+    ///
+    /// ## Differences between Base and Altair
+    ///
+    /// - Base: any attestation can match the head.
+    /// - Altair: only "timely" attestations can match the head.
+    pub fn previous_epoch_head_attesting_balance(&self) -> Result<u64, ParticipationCacheError> {
+        match self {
+            EpochProcessingSummary::Base { total_balances, .. } => {
+                Ok(total_balances.previous_epoch_head_attesters())
+            }
+            EpochProcessingSummary::Altair {
+                participation_cache,
+            } => participation_cache.previous_epoch_head_attesting_balance(),
+        }
+    }
+
+    /// Returns the sum of the effective balance of all validators in the previous epoch who
+    /// included an attestation that matched the source.
+    ///
+    /// ## Differences between Base and Altair
+    ///
+    /// - Base: any attestation can match the source.
+    /// - Altair: only "timely" attestations can match the source.
+    pub fn previous_epoch_source_attesting_balance(&self) -> Result<u64, ParticipationCacheError> {
+        match self {
+            EpochProcessingSummary::Base { total_balances, .. } => {
+                Ok(total_balances.previous_epoch_attesters())
+            }
+            EpochProcessingSummary::Altair {
+                participation_cache,
+            } => participation_cache.previous_epoch_source_attesting_balance(),
+        }
+    }
+
+    /// Returns `true` if `val_index` was included in the active validator indices in the previous
+    /// epoch *and* the validator is not slashed.
+    ///
+    /// ## Notes
+    ///
+    /// Always returns `false` for an unknown `val_index`.
+    pub fn is_active_unslashed_in_previous_epoch(&self, val_index: usize) -> bool {
+        match self {
+            EpochProcessingSummary::Base { statuses, .. } => statuses
+                .get(val_index)
+                .map_or(false, |s| s.is_active_in_previous_epoch && s.is_slashed),
```

Comment:
Missing a negation on `s.is_slashed`

```suggestion
                .map_or(false, |s| s.is_active_in_previous_epoch && !s.is_slashed),
```

Reply:
Good catch, thank you :pray: